<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>TrajConserve Workflow • TrajConserve</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Roboto-0.4.9/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.9/font.css" rel="stylesheet">
<link href="../deps/Roboto_Slab-0.4.9/font.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="TrajConserve Workflow">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TrajConserve</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/introduction.html">Introduction</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/conservation_analysis.html">Conservation Analysis with TrajConserve</a></li>
    <li><a class="dropdown-item" href="../articles/introduction.html">Introduction to TrajConserve</a></li>
    <li><a class="dropdown-item" href="../articles/stan_note.html">Note About Stan Dependencies</a></li>
    <li><a class="dropdown-item" href="../articles/workflow.html">TrajConserve Workflow</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/GilbertHan1011/TrajConserve" aria-label="GitHub"><span class="fa fab fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>TrajConserve Workflow</h1>
                        <h4 data-toc-skip class="author">Gilbert
Han</h4>
            
            <h4 data-toc-skip class="date">2025-03-28</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/GilbertHan1011/TrajConserve/blob/main/vignettes/workflow.Rmd" class="external-link"><code>vignettes/workflow.Rmd</code></a></small>
      <div class="d-none name"><code>workflow.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette demonstrates a complete workflow for trajectory
conservation analysis with the TrajConserve package. TrajConserve
enables the identification of conserved and non-conserved gene
expression patterns across developmental trajectories in single-cell
RNA-seq data.</p>
<p>The workflow includes:</p>
<ol style="list-style-type: decimal">
<li>Loading and preparing data</li>
<li>Converting Seurat objects to trajectory arrays</li>
<li>Running Bayesian GAM models on gene expression trajectories</li>
<li>Storing results in HDF5 files for efficiency</li>
<li>Extracting and visualizing model metrics</li>
<li>Calculating conservation scores to identify conserved genes</li>
</ol>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Before starting, make sure you have TrajConserve installed:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install from GitHub</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"GilbertHan1011/TrajConserve"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install required dependencies</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"rhdf5"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"brms"</span>, <span class="st">"ggplot2"</span>, <span class="st">"pheatmap"</span>, <span class="st">"ggrepel"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="setting-up-stan-important">Setting up Stan (Important!)<a class="anchor" aria-label="anchor" href="#setting-up-stan-important"></a>
</h3>
<p>TrajConserve relies on Stan for Bayesian modeling. By default, it
uses the rstan backend, but for optimal performance, we strongly
recommend installing and configuring cmdstanr:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install cmdstanr from Stan R-universe (recommended method)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"cmdstanr"</span>, repos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'https://stan-dev.r-universe.dev'</span>, <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"repos"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu">cmdstanr</span><span class="fu">::</span><span class="fu">check_cmdstan_toolchain</span><span class="op">(</span>fix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fl">600</span><span class="op">)</span>  <span class="co"># Set a longer timeout for downloading</span></span>
<span><span class="fu">cmdstanr</span><span class="fu">::</span><span class="fu">install_cmdstan</span><span class="op">(</span>cores <span class="op">=</span> <span class="fl">2</span>,overwrite<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>These settings ensure that Stan can properly compile with your C++
toolchain. After installing cmdstanr, you can use it by setting
<code>backend = "cmdstanr"</code> in the modeling functions. Without
these configurations, you might encounter errors like:</p>
</div>
</div>
<div class="section level2">
<h2 id="load-required-packages">Load Required Packages<a class="anchor" aria-label="anchor" href="#load-required-packages"></a>
</h2>
<p>Start by loading the necessary packages:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/GilbertHan1011/TrajConserve" class="external-link">TrajConserve</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/grimbough/rhdf5" class="external-link">rhdf5</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-and-prepare-data">Load and Prepare Data<a class="anchor" aria-label="anchor" href="#load-and-prepare-data"></a>
</h2>
<p>First, we load a prepared Seurat object containing single-cell data
with pseudotime information:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load example data from the package</span></span>
<span><span class="va">small_epi</span> <span class="op">&lt;-</span> <span class="fu">TrajConserve</span><span class="fu">::</span><span class="fu">load_example_data</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Examine the structure</span></span>
<span><span class="va">small_epi</span></span></code></pre></div>
<p>The Seurat object should contain: - Expression data in an appropriate
assay - Pseudotime values in metadata - Project/sample/batch information
in metadata</p>
</div>
<div class="section level2">
<h2 id="convert-seurat-object-to-trajectory-array">Convert Seurat Object to Trajectory Array<a class="anchor" aria-label="anchor" href="#convert-seurat-object-to-trajectory-array"></a>
</h2>
<p>Next, we convert the Seurat object to a format suitable for
trajectory modeling:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trajectory_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/seurat_to_trajectory_array.html">seurat_to_trajectory_array</a></span><span class="op">(</span></span>
<span>  seurat_obj <span class="op">=</span> <span class="va">small_epi</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"originalexp"</span>,      <span class="co"># Assay containing expression data</span></span>
<span>  pseudo_col <span class="op">=</span> <span class="st">"pseudo"</span>,      <span class="co"># Column with pseudotime values</span></span>
<span>  project_col <span class="op">=</span> <span class="st">"Project"</span>     <span class="co"># Column with project/sample identifiers</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Examine the structure of the trajectory data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">trajectory_data</span><span class="op">)</span></span></code></pre></div>
<p>The <code>seurat_to_trajectory_array</code> function creates: - A 3D
array of expression data organized by gene, pseudotime, and sample -
Gene names and sample metadata - Binned pseudotime values for
modeling</p>
</div>
<div class="section level2">
<h2 id="test-a-single-trajectory-model">Test a Single Trajectory Model<a class="anchor" aria-label="anchor" href="#test-a-single-trajectory-model"></a>
</h2>
<p>Before running models on multiple genes, it’s often helpful to test
with a single gene:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Select a gene index to test</span></span>
<span><span class="va">gene_idx</span> <span class="op">&lt;-</span> <span class="fl">10</span>  <span class="co"># Example gene index</span></span>
<span></span>
<span><span class="co"># Run a test model for one gene</span></span>
<span><span class="va">model_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_trajectory_model.html">run_trajectory_model</a></span><span class="op">(</span><span class="va">trajectory_data</span><span class="op">$</span><span class="va">reshaped_data</span>, <span class="va">gene_idx</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize the model results</span></span>
<span><span class="fu"><a href="../reference/plot_results_brms.html">plot_results_brms</a></span><span class="op">(</span><span class="va">model_test</span><span class="op">)</span></span></code></pre></div>
<p>This creates a test model for one gene and visualizes the trajectory
across samples. The plot shows the fitted GAM curve with uncertainty
intervals for each sample.</p>
</div>
<div class="section level2">
<h2 id="run-multiple-models-and-save-results">Run Multiple Models and Save Results<a class="anchor" aria-label="anchor" href="#run-multiple-models-and-save-results"></a>
</h2>
<p>For comprehensive analysis, we run models for multiple genes in
parallel:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define output directory</span></span>
<span><span class="va">savedir</span> <span class="op">&lt;-</span> <span class="st">"./"</span></span>
<span></span>
<span><span class="co"># Run models for multiple genes</span></span>
<span><span class="fu">TrajConserve</span><span class="fu">::</span><span class="fu"><a href="../reference/run_multiple_models.html">run_multiple_models</a></span><span class="op">(</span></span>
<span>  <span class="va">trajectory_data</span><span class="op">$</span><span class="va">reshaped_data</span>,</span>
<span>  gene_indices <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>,      <span class="co"># First 20 genes (adjust as needed)</span></span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  n_cores <span class="op">=</span> <span class="fl">4</span>,              <span class="co"># Adjust based on your computing resources</span></span>
<span>  backend <span class="op">=</span> <span class="st">"rstan"</span>,        <span class="co"># cmdstanr is highly recomendded</span></span>
<span>  save_metrics <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  save_metrics_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">savedir</span>, <span class="st">"test1.hdf5"</span><span class="op">)</span>,</span>
<span>  save_plots <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  save_plots_dir <span class="op">=</span> <span class="va">savedir</span>,</span>
<span>  save_models <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  save_models_dir <span class="op">=</span> <span class="va">savedir</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This process: - Runs Bayesian GAM models for each selected gene -
Executes in parallel for efficiency - Saves model metrics to an HDF5
file - Saves diagnostic plots as PDFs - Saves the model objects as RDS
files</p>
</div>
<div class="section level2">
<h2 id="access-and-examine-hdf5-results">Access and Examine HDF5 Results<a class="anchor" aria-label="anchor" href="#access-and-examine-hdf5-results"></a>
</h2>
<p>After model fitting, we can examine the saved results:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Open the HDF5 file</span></span>
<span><span class="va">h5_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">savedir</span>, <span class="st">"test1.hdf5"</span><span class="op">)</span></span>
<span><span class="va">h5_test</span> <span class="op">&lt;-</span> <span class="fu">rhdf5</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/H5Fopen.html" class="external-link">H5Fopen</a></span><span class="op">(</span><span class="va">h5_file</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># List the contents of the HDF5 file</span></span>
<span><span class="fu">rhdf5</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/h5ls.html" class="external-link">h5ls</a></span><span class="op">(</span><span class="va">h5_file</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Examine the weight matrix</span></span>
<span><span class="va">h5_test</span><span class="op">$</span><span class="st">"array_weights"</span></span>
<span></span>
<span><span class="co"># Close the HDF5 file</span></span>
<span><span class="fu">rhdf5</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/H5Fclose.html" class="external-link">H5Fclose</a></span><span class="op">(</span><span class="va">h5_test</span><span class="op">)</span></span></code></pre></div>
<p>The HDF5 file contains various metrics from our models, including: -
Coefficient estimates - Standard errors - R-squared values -
Significance measures - And more</p>
</div>
<div class="section level2">
<h2 id="extract-metrics-and-visualize-results">Extract Metrics and Visualize Results<a class="anchor" aria-label="anchor" href="#extract-metrics-and-visualize-results"></a>
</h2>
<p>TrajConserve provides functions to extract metrics and visualize
results:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract metrics from the HDF5 file</span></span>
<span><span class="va">metric</span> <span class="op">&lt;-</span> <span class="fu">TrajConserve</span><span class="fu">::</span><span class="fu"><a href="../reference/extract_hdf5_metric.html">extract_hdf5_metric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">savedir</span>,<span class="st">"test1.hdf5"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Examine the extracted metrics</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">metric</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">metric</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a heatmap visualization of the metrics</span></span>
<span><span class="fu">TrajConserve</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_hdf5_heatmap.html">plot_hdf5_heatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">savedir</span>, <span class="st">"test1.hdf5"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The <code>extract_hdf5_metric</code> function retrieves a specific
metric from the HDF5 file, while <code>plot_hdf5_heatmap</code> creates
a visualization showing patterns across genes and samples.</p>
</div>
<div class="section level2">
<h2 id="calculate-conservation-scores">Calculate Conservation Scores<a class="anchor" aria-label="anchor" href="#calculate-conservation-scores"></a>
</h2>
<p>Finally, we calculate conservation scores to identify conserved
genes:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate conservation scores</span></span>
<span><span class="va">conservation_results</span> <span class="op">&lt;-</span> <span class="fu">TrajConserve</span><span class="fu">::</span><span class="fu"><a href="../reference/calculate_conservation.html">calculate_conservation</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">savedir</span>,<span class="st">"test1.hdf5"</span><span class="op">)</span>,</span>
<span>  metric <span class="op">=</span> <span class="st">"Estimate"</span>,</span>
<span>  mean_weight <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  variability_weight <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>  conservation_threshold <span class="op">=</span> <span class="fl">0.7</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Examine conservation results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">conservation_results</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Count conserved and non-conserved genes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">conservation_results</span><span class="op">$</span><span class="va">is_conserved</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create visualizations of conservation results</span></span>
<span><span class="fu"><a href="../reference/plot_conservation.html">plot_conservation</a></span><span class="op">(</span></span>
<span>  <span class="va">conservation_results</span>,</span>
<span>  plot_type <span class="op">=</span> <span class="st">"scatter"</span>,</span>
<span>  highlight_n <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  file_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">savedir</span>, <span class="st">"conservation_scatter.pdf"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_conservation.html">plot_conservation</a></span><span class="op">(</span></span>
<span>  <span class="va">conservation_results</span>,</span>
<span>  plot_type <span class="op">=</span> <span class="st">"histogram"</span>,</span>
<span>  file_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">savedir</span>,<span class="st">"conservation_histogram.pdf"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The conservation analysis: - Calculates a conservation score based on
mean expression and variability - Classifies genes as conserved or
non-conserved - Provides visualizations to interpret the results</p>
</div>
<div class="section level2">
<h2 id="advanced-visualization">Advanced Visualization<a class="anchor" aria-label="anchor" href="#advanced-visualization"></a>
</h2>
<p>For publication-quality visualizations, we can combine conservation
results with expression patterns:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a heatmap with gene conservation annotation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract estimate matrix</span></span>
<span><span class="va">estimate_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_hdf5_metric.html">extract_hdf5_metric</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">savedir</span>,<span class="st">"test1.hdf5"</span><span class="op">)</span>, <span class="st">"Estimate"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create annotation data frame</span></span>
<span><span class="va">gene_type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">conservation_results</span><span class="op">$</span><span class="va">is_conserved</span>, <span class="st">"Conserved"</span>, <span class="st">"Non-conserved"</span><span class="op">)</span></span>
<span><span class="va">gene_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  Conservation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">gene_type</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Conserved"</span>, <span class="st">"Non-conserved"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  row.names <span class="op">=</span> <span class="va">conservation_results</span><span class="op">$</span><span class="va">gene</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create heatmap</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">estimate_matrix</span><span class="op">)</span>, </span>
<span>         annotation_row <span class="op">=</span> <span class="va">gene_anno</span>,</span>
<span>         main <span class="op">=</span> <span class="st">"Expression Patterns of Conserved vs Non-conserved Genes"</span>,</span>
<span>         filename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"conservation_heatmap.pdf"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This creates a heatmap that shows expression patterns across samples,
with genes annotated by their conservation status.</p>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>This workflow demonstrates the complete process of trajectory
conservation analysis with TrajConserve:</p>
<ol style="list-style-type: decimal">
<li>We prepared single-cell data with pseudotime information</li>
<li>Converted it to trajectory arrays suitable for modeling</li>
<li>Ran Bayesian GAM models to characterize expression trajectories</li>
<li>Stored results efficiently in HDF5 files</li>
<li>Extracted metrics and created visualizations</li>
<li>Calculated conservation scores to identify conserved genes</li>
</ol>
<p>The identified conserved genes represent expression patterns that are
maintained across samples/conditions/species, potentially indicating
functionally important developmental programs.</p>
</div>
<div class="section level2">
<h2 id="next-steps">Next Steps<a class="anchor" aria-label="anchor" href="#next-steps"></a>
</h2>
<p>After identifying conserved and non-conserved genes, you might:</p>
<ul>
<li>Perform Gene Ontology enrichment analysis on each gene set</li>
<li>Integrate with other types of genomic data</li>
<li>Compare conservation across different developmental stages</li>
<li>Investigate specific gene modules of interest in greater detail</li>
</ul>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gilbert Han.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
